api/envoy/service/health/v3/hds.proto

```diff
@@ -108,11 +108,30 @@ message EndpointHealth {
   config.core.v3.HealthStatus health_status = 2;
 }
 
+// Group endpoint health by locality under each cluster.
+message LocalityEndpointsHealth {
+  api.v3.core.Locality locality = 1;
+
+  repeated EndpointHealth endpoints_health = 2;
+}
+
+// The health status of endpoints in a cluster. The cluster name and locality
+// should match the corresponding fields in ClusterHealthCheck message.
+message ClusterEndpointsHealth {
+  string cluster_name = 1;
+
+  repeated LocalityEndpointsHealth locality_endpoints_health = 2;
+}
+
 message EndpointHealthResponse {
   option (udpa.annotations.versioning).previous_message_type =
       "envoy.service.discovery.v2.EndpointHealthResponse";
 
+  // Flat list of endpoint health information.
```

Comment:
Maybe better to mention this field will be deprecated, and a TODO to change implementation to use cluster_endpoints_health instead.

Reply:
Yep, I suggest both deprecating this and also using `(udpa.annotations.field_migrate).oneof_promotion` to indicate that this is a future `oneof`.
/wait

Reply:
Sorry, I don't quite get the purpose of the future oneof.
I thought the field "endpoints_health" will be replaced by cluster_endpoints_health.
In this change, we add the new field, with some comment to mention the old field will be removed. And later after moving to the new field in production code, we will remove the old field.

Reply:
@xdzhai From what I understand this change is to be backwards-compatible to remain in v3 according to the [API versioning](https://github.com/envoyproxy/envoy/blob/master/api/API_VERSIONING.md) guidelines, else it would need to wait until v4.

Reply:
Thanks for sharing the link!
In v3 proto, shall we mark the old field as [deprecated = true]? Maybe add a TODO first, then mark "[deprecated = true]" after finishing code implementation.
In v4alpha proto, may we remove the old field directly?

Reply:
@htuch  I'm also unsure about this. 

If both 
`deprecated = true, (udpa.annotations.field_migrate).oneof_promotion = "endpoints_health_result" ] `
is used on the field, then the deprecation removes it from v4 so the **oneof** is now missing a field and has one entry, making the oneof redundant. Which of the two should I use in this case? Should it be marked as deprecated in docs only and use oneof, or just deprecated = true?

I'm unsure what the implementation will do when it tries to write to the backwards-compatible field that doesn't exist once v4 is released. In my implementation, should I be supporting this field by continuing to populate it as well as the new one, or stop populating it and only fill out the new one? 

If I do the former than this will break the oneof, and only recognize the "deprecated" field as the payload since both are present. If we go with the deprecation annotation approach, then when v4 is released the implementation will try to write to a field that doesn't exist (assuming I am to be populating the old field in v3 for backwards-compatibility).

If I do the latter and stop populating the old field, than the backwards-compatible field isn't really backwards-compatible.

I understand how data annotations are used to generate v4alpha by protoxtool, but what kinds of deprecation standards are to be followed in implementation? I could not find much information for this area after looking through the code and docs.

Thanks!

Reply:
That's a fair point. Since it's deprecated and we have a new one, maybe let's just document that if you specify the more general locality endpoints, it is consumed preferentially.

What I recommend in the implementation is that you internally translate from the old to the new field early in the config ingestion.

Reply:
I guess this EndpointHealthResponse is only consumed by customers' code.
To be backwards compatible, if there may be any other user reading from the old field, we might have to populate both fields?